src/ppm.rs

```diff
@@ -0,0 +1,499 @@
+    V: pcp::Value<F>,
+{
+    input_share: V,
+    verifier_share: Vec<F>,
+    joint_rand_seed_hint: Vec<u8>,
+}
+
+/// Run by each aggregator, this consumes the [`Upload`] message sent from the client and
+/// produces the [`Verifier`] message that will be broadcast to the other aggregators. It also
+/// returns the aggregator's state.
+///
+/// # Parameters
+///
+/// * `msg` is the message sent by the client. It contains, among other things, an input and proof
+/// share.
+/// * `param` is used to reconstruct the input share from the raw message.
+/// * `query_rand_seed` is used to derive the query randomness shared by all the aggregators.
+pub fn verify_start<F, V>(
+    msg: Upload<F>,
+    param: V::Param,
+    aggregator_id: u8,
+    query_rand_seed: &[u8],
+) -> Result<(AggregatorState<F, V>, Verifier<F>), PpmError>
+where
+    F: FieldElement,
+    V: pcp::Value<F>,
+{
+    assert_safe();
+
+    let input_share_data: Vec<F> = Vec::try_from(msg.input_share)?;
+    let input_share = V::try_from((param, &input_share_data))?;
+
+    let proof_share_data: Vec<F> = Vec::try_from(msg.proof_share)?;
+    let proof_share = pcp::Proof::from(proof_share_data);
+
+    // Compute the joint randomness.
+    let mut joint_rand_seed_share = vec![0; SEED_LENGTH];
+    let mut hasher = blake3::Hasher::new();
+    hasher.update(&[aggregator_id]);
+    hasher.update(&msg.blind);
+    for x in input_share.as_slice() {
+        hasher.update(&(*x).into());
+    }
+    joint_rand_seed_share.copy_from_slice(hasher.finalize().as_bytes());
+
+    let mut joint_rand_seed = vec![0; SEED_LENGTH];
+    for j in 0..SEED_LENGTH {
+        joint_rand_seed[j] = msg.joint_rand_seed_hint[j] ^ joint_rand_seed_share[j];
+    }
+
+    let mut prng = Prng::try_from_seed(&joint_rand_seed)?;
+    let joint_rand: Vec<F> = (0..input_share.joint_rand_len())
+        .map(|_| prng.next().unwrap())
+        .collect();
+
+    // Compute the query randomness.
+    let mut prng = Prng::try_from_seed(query_rand_seed)?;
+    let query_rand: Vec<F> = (0..input_share.query_rand_len())
+        .map(|_| prng.next().unwrap())
+        .collect();
+
+    // Run the query-generation algorithm.
+    let verifier_share = query(&input_share, &proof_share, &query_rand, &joint_rand)?;
+    let verifier_share_data = verifier_share.as_slice().to_vec();
+
+    // Prepare the output state and message.
+    let state = AggregatorState {
+        input_share,
+        verifier_share: verifier_share_data.clone(),
+        joint_rand_seed_hint: msg.joint_rand_seed_hint,
+    };
+
+    let out_msg = Verifier {
+        verifier_share: verifier_share_data,
+        joint_rand_seed_share,
+    };
+
+    Ok((state, out_msg))
+}
+
+/// Run by each aggregator, this consumes the [`Verifier`] messages broadcast by each of the
+/// other aggregators. It returns the aggregator's input share only if the input is valid.
+pub fn verify_finish<F, V>(
+    state: AggregatorState<F, V>,
+    msgs: Vec<Verifier<F>>,
```

Comment:
```suggestion
    verifier_messages: Vec<Verifier<F>>,
```

Reply:
Again, I have a weak preference for a short, generic name here.